common/nym-kcp/src/session.rs

```diff
@@ -0,0 +1,1778 @@
+            self.conv,
+            thread::current().id(),
+            una,
+            removed_count,
+            original_len,
+            self.snd_buf.len(),
+            post_retain_sns
+        );
+
+        if removed_count > 0 {
+            // Use trace level if no segments were removed but buffer wasn't empty
+            debug!(
+                "[ConvID: {}, Thread: {:?}] parse_una(una={}): No segments removed from snd_buf (len={}). Remaining sns: {:?}",
+                self.conv,
+                thread::current().id(),
+                una,
+                original_len,
+                self.snd_buf.iter().map(|s| s.sn).collect::<Vec<_>>()
+            );
+        }
+
+        // Update the known acknowledged sequence number.
+        // Use max to prevent out-of-order packets with older UNA from moving snd_una backwards.
+        self.snd_una = cmp::max(self.snd_una, una);
+    }
+
+    fn parse_ack(&mut self, sn: u32, ts: u32) {
+        debug!(
+            "[ConvID: {}, Thread: {:?}] Parsing ack, sn: {}, ts: {}",
+            self.conv,
+            thread::current().id(),
+            sn,
+            ts
+        );
+        // find the segment in snd_buf
+        if let Some(pos) = self.snd_buf.iter().position(|seg| seg.sn == sn) {
+            let seg = self.snd_buf.remove(pos).unwrap();
+            debug!(
+                "[ConvID: {}, Thread: {:?}] Acked segment, sn: {}, frg: {}",
+                self.conv,
+                thread::current().id(),
+                sn,
+                seg.frg
+            );
+            // update RTT
+            let rtt = self.current.saturating_sub(ts);
+            self.update_rtt(rtt);
+        } else {
+            // Log if the segment was NOT found
+            let current_sns: Vec<u32> = self.snd_buf.iter().map(|s| s.sn).collect();
+            warn!(
+                "[ConvID: {}, Thread: {:?}] parse_ack: ACK received for sn={}, but segment not found in snd_buf (len={}): {:?}",
+                self.conv,
+                thread::current().id(),
+                sn,
+                self.snd_buf.len(),
+                current_sns
+            );
+        }
+    }
+
+    fn parse_data(&mut self, pkt: &KcpPacket) {
+        // Insert into rcv_buf if pkt.sn in [rcv_nxt .. rcv_nxt + rcv_wnd)
+        if pkt.sn() >= self.rcv_nxt + self.rcv_wnd as u32 {
+            // out of window
+            return;
+        }
+        if pkt.sn() < self.rcv_nxt {
+            // already got it, discard
+            return;
+        }
+
+        // Check if we have it
+        let mut insert_idx = self.rcv_buf.len();
+        for (i, seg) in self.rcv_buf.iter().enumerate() {
+            #[allow(clippy::comparison_chain)]
+            if pkt.sn() < seg.sn {
+                insert_idx = i;
+                break;
+            } else if pkt.sn() == seg.sn {
+                // duplicate
+                return;
```

Comment:
while this seems to be a sensible choice, if you receive a duplicate packet you won't be checking your buffer, i.e. `self.move_buf_to_queue();` meanwhile in the C implementation (which I assume you based this on) , they still check the queue regardless of the packet being repeated